agent/fw_drupal8.c

```diff
@@ -605,12 +605,128 @@ NR_PHP_WRAPPER(nr_drupal94_invoke_all_with_clean) {
 NR_PHP_WRAPPER_END
 #endif  // OAPI
 
+#define NR_FREE_HOOK_MEM \
+  nr_free(hook_str);     \
+  nr_free(class_str);    \
+  nr_free(method_str);   \
+  nr_free(module_str);   \
+  nr_free(hookpath);
+
+/*
+ * Purpose: Instrument Drupal Attribute Hooks for Drupal 11.1+
+ *
+ * Params: 1. A zval pointer to the moduleHandler instance in use by Drupal.
+ *
+ * Return: bool
+ *
+ */
+static bool nr_drupal_hook_attribute_instrument(zval* module_handler) {
+  zval* hook_implementation_map = NULL;
+
+  nr_php_string_hash_key_t* hook_key = NULL;
+  zval* hook_val = NULL;
+  nr_php_string_hash_key_t* class_key = NULL;
+  zval* class_val = NULL;
+  nr_php_string_hash_key_t* method_key = NULL;
+  zval* module_val = NULL;
+  zend_ulong key_num = 0;
+
+  char* hook_str = NULL;
+  char* class_str = NULL;
+  char* method_str = NULL;
+  char* module_str = NULL;
+  char* hookpath = NULL;
+
+  hook_implementation_map = nr_php_get_zval_object_property(
+      module_handler, "hookImplementationsMap");
+  if (hook_implementation_map) {
+    if (nr_php_is_zval_valid_array(hook_implementation_map)) {
+      ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(hook_implementation_map), key_num,
+                                hook_key, hook_val) {
+        (void)key_num;
+        if ((NULL == hook_key) || (0 == nr_php_is_zval_valid_array(hook_val))) {
+          nrl_warning(NRL_FRAMEWORK,
+                      "hookImplementationsMap[hook = %s]: invalid value",
+                      NRSAFESTR(ZEND_STRING_VALUE(hook_key)));
+          NR_FREE_HOOK_MEM
+          return false;
+        }
+
+        ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(hook_val), key_num, class_key,
+                                  class_val) {
+          (void)key_num;
+          if ((NULL == class_key)
+              || (0 == nr_php_is_zval_valid_array(class_val))) {
+            nrl_warning(NRL_FRAMEWORK,
+                        "hookImplementationsMap[class = %s]: invalid value",
+                        NRSAFESTR(ZEND_STRING_VALUE(class_key)));
+            NR_FREE_HOOK_MEM
+            return false;
+          }
+
+          ZEND_HASH_FOREACH_KEY_VAL(Z_ARRVAL_P(class_val), key_num, method_key,
+                                    module_val) {
+            (void)key_num;
+
+            NR_FREE_HOOK_MEM
+
+            if ((NULL == method_key)
+                || (0 == nr_php_is_zval_valid_string(module_val))) {
+              nrl_warning(NRL_FRAMEWORK,
+                          "hookImplementationsMap[method = %s]: invalid value",
+                          NRSAFESTR(ZEND_STRING_VALUE(method_key)));
+              NR_FREE_HOOK_MEM
+              return false;
+            }
+
+            hook_str = nr_strdup(ZEND_STRING_VALUE(hook_key));
+            class_str = nr_strdup(ZEND_STRING_VALUE(class_key));
+            method_str = nr_strdup(ZEND_STRING_VALUE(method_key));
+            module_str = nr_strdup(Z_STRVAL_P(module_val));
```

Comment:
There is no need for any of these `strdup`s:
- `ZEND_STRING_VALUE(class_key)` and `ZEND_STRING_VALUE(method_key)` can be simply passed as arguments to `nr_formatf` that construct `hookpath`
- `Z_STRVAL_P(module_val)` and `ZEND_STRING_VALUE(hook_key)` can be simply passed as arguments to `nr_php_wrap_user_function_drupal` which does `strdup`. As an added benefit, zend_strings have length associated with them so there's no need to call strlen for neither `module_val` nor `hook_key`.

Reply:
With the suggestion above, there's really no need for `NR_FREE_HOOK_MEM` macro to exist as well.

Reply:
Addressed here 3ded06d732c7cd7b4ee87b62f4d57a3dbbc60ef8